rfcs/0008-transition-frontier-controller.md

```diff
@@ -0,0 +1,337 @@
+
+It's two orders of magnitude more computationally expensive to constantly coalesce masked merkle databases in the worst case than to just keep them separated (back of napkin math). Moreover, we are able to in $O(1)$ time answer sync ledger queries if we have ledger builders at every position.
+
+<a href="ktree"></a>
+#### Ktree
+
+An `Ktree.t` is a rose-tree with a fixed-depth. The new `ktree` will support the following $O(1)$ operations:
+
+```ocaml
+(* assuming a key_of_value : value -> key function is available *)
+lookup : key -> value
+add : value -> parent:key -> [`No_parent | `Added | `Repeat]
+```
+
+This is acheived by using a `(key, value node) Hash table` and a rose tree with
+a mutable children field. A `lookup` just checks the hash table. An `add`
+looks up `parent` in the hash table, if present it adds a child to that node.
+
+For now, we will wrap the existing ktree up in this signature and the
+operations will not be $O(1)$. If integration tests need this extra performance
+boost, we will then build this component.
+
+<a href="ledger-catchup"></a>
+### Ledger Catchup
+
+Input: `External_transition.t` (from catchup monitor)
+Output: `Breadcrumb.t` (to processor)
+
+Ledger catchup runs a single catchup worker job at a time. Whenever catchup
+monitor descides it's time for a new catchup job to start, it will send
+something on the pipe.
+
+#### Catchup worker
+
+A worker is responsible for performing a history sync.
+
+In the future, we will support multiple catchup workers at once. See [unresolved questions](#unresolved-questions).
+
+New jobs handed to the same worker cause a retargeting action where we don't
+throw out existing data, but start again getting data from the source.
+
+<a href="history-sync"></a>
+#### History Sync
+
+History sync is a new process that is not yet implemented in the current
+ledger-builder-controller.
+
+`history_sync transition` asynchronously walks backwards downloading each
+external transition until either (1) it reconnects with some existing breadcrumb
+or(2) it passes the locked slot without reconnecting. In either case, we then
+walk forwards doing `Path_traversal` logic materializing masked ledger-builders
+all the way up the path. The root breadcrumb then gets dispatched back into the
+processor.
+
+The details of this download process are TBD. It will likely make sense to ask
+for a manifest of locations for the transitions all at once and to do some sort
+of torrenting solution. In the short term, we'll just naively download one
+transition at a time.
+
+If we do pass the locked slot without reconnecting, we need to perform an
+additional step of invoking the sync ledger process.
+
+<a href="sync-handler"></a>
+### Sync Handler
```

Comment:
I think handling both sync queries and history queries makes sense. However, I think that we should rename this then from `Sync_handler` to something like `Query_handler`.